File: src/bot/commands/ytsearch.ts
import type { Message } from 'discord.js-selfbot-v13';
import { spawn } from 'child_process';

interface YouTubeSearchResult {
  id: string;
  title: string;
  duration: string;
  channel: string;
  url: string;
}

const searchCache = new Map<string, { results: YouTubeSearchResult[], timestamp: number }>();
const CACHE_DURATION = 5 * 60 * 1000; // 5 minutes

async function searchYouTube(query: string, limit = 10): Promise<YouTubeSearchResult[]> {
  return new Promise((resolve) => {
    const ytdlp = spawn('yt-dlp', [
      '--dump-json',
      '--flat-playlist',
      '--no-warnings',
      '-I', `1:${limit}`,
      `ytsearch${limit}:${query}`
    ]);

    let output = '';
    let error = '';

    ytdlp.stdout.on('data', (data) => {
      output += data.toString();
    });

    ytdlp.stderr.on('data', (data) => {
      error += data.toString();
    });

    ytdlp.on('close', (code) => {
      if (code !== 0 || !output) {
        console.error('[YouTubeSearch] yt-dlp error:', error);
        resolve([]);
        return;
      }

      try {
        const results: YouTubeSearchResult[] = [];
        const lines = output.trim().split('\n');
        
        for (const line of lines) {
          if (!line.trim()) continue;
          const info = JSON.parse(line);
          results.push({
            id: info.id,
            title: info.title || 'Unknown',
            duration: formatDuration(info.duration),
            channel: info.channel || info.uploader || 'Unknown',
            url: info.url || `https://www.youtube.com/watch?v=${info.id}`,
          });
        }
        
        resolve(results);
      } catch (e) {
        console.error('[YouTubeSearch] Failed to parse results:', e);
        resolve([]);
      }
    });

    ytdlp.on('error', (err) => {
      console.error('[YouTubeSearch] yt-dlp spawn error:', err);
      resolve([]);
    });
  });
}

function formatDuration(seconds: number | null): string {
  if (!seconds) return 'Live/Unknown';
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  const s = seconds % 60;
  if (h > 0) {
    return `${h}:${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`;
  }
  return `${m}:${String(s).padStart(2, '0')}`;
}

export function getYouTubeSearchResult(userId: string, index: number): YouTubeSearchResult | null {
  const cached = searchCache.get(userId);
  if (!cached || Date.now() - cached.timestamp > CACHE_DURATION) {
    return null;
  }
  return cached.results[index - 1] || null;
}

export async function ytSearchCommand(message: Message, args: string[]): Promise<void> {
  if (!message.guild) {
    await message.channel.send('❌ This command can only be used in a server');
    return;
  }

  const query = args.join(' ');
  if (!query) {
    await message.channel.send(
      '❌ Usage: `!yts <search query>`\n' +
      'Example: `!yts never gonna give you up`\n\n' +
      'Then use `!ytp <number>` to play a result'
    );
    return;
  }

  const statusMsg = await message.channel.send(`🔍 Searching YouTube for "${query}"...`);

  try {
    const results = await searchYouTube(query, 10);

    if (results.length === 0) {
      await statusMsg.edit('❌ No results found. Try a different search query.');
      return;
    }

    // Cache results
    searchCache.set(message.author.id, {
      results,
      timestamp: Date.now(),
    });

    const resultLines = results.map((r, i) => 
      `**${i + 1}.** ${r.title}\n   └ ${r.channel} • ${r.duration}`
    );

    await statusMsg.edit(
      `🎬 **YouTube Search Results:**\n\n` +
      resultLines.join('\n\n') +
      `\n\n*Use \`!ytp <number>\` to play a result*`
    );
  } catch (error) {
    console.error('[YouTubeSearch] Error:', error);
    await statusMsg.edit(`❌ Search failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
}

export async function ytPlayCommand(message: Message, args: string[]): Promise<void> {
  if (!message.guild) {
    await message.channel.send('❌ This command can only be used in a server');
    return;
  }

  const selection = parseInt(args[0], 10);
  if (isNaN(selection) || selection < 1) {
    await message.channel.send('❌ Usage: `!ytp <number>` (use `!yts` to search first)');
    return;
  }

  const result = getYouTubeSearchResult(message.author.id, selection);
  if (!result) {
    await message.channel.send('❌ Invalid selection or search expired. Use `!yts` to search first.');
    return;
  }

  // Import and call youtubeCommand with the URL
  const { youtubeCommand } = await import('./youtube.js');
  await youtubeCommand(message, [result.url]);
}

File: selfbot/commands/ytsearch.py
"""YouTube search and play commands via IPC."""

from __future__ import annotations

from typing import TYPE_CHECKING
import discord

from ..ipc import ipc
from ..helpers import get_voice_channel, format_duration

if TYPE_CHECKING:
    from ..client import SchroStreamBot


async def yt_search_command(bot: SchroStreamBot, message: discord.Message, args: list[str]) -> None:
    query = " ".join(args)
    if not query:
        await message.channel.send("❌ Usage: `!yts <query>`")
        return

    status_msg = await message.channel.send(f'🔍 Searching YouTube for "{query}"...')

    result = await ipc.youtube_search(query)

    if not result.get("success"):
        await status_msg.edit(content=f"❌ {result.get('error', 'Search failed')}")
        return

    results = result.get("results", [])
    if not results:
        await status_msg.edit(content=f'❌ No results found for "{query}"')
        return

    # Cache for yt play command
    bot.yt_search_sessions[str(message.author.id)] = results

    lines: list[str] = []
    for i, item in enumerate(results[:20], 1):
        title = item.get("title", "Unknown")
        duration = item.get("duration", "N/A")
        channel = item.get("channel", "Unknown")
        views = item.get("views", "")
        views_str = f" • {views}" if views else ""
        lines.append(f"**{i}.** {title} [{duration}]\n   *{channel}{views_str}*")

    await status_msg.edit(
        content=f'🎬 **YouTube Results for "{query}":**\n\n' + "\n".join(lines) + "\n\n*Use `!ytp <number>` to play*"
    )


async def yt_play_command(bot: SchroStreamBot, message: discord.Message, args: list[str]) -> None:
    if not message.guild:
        await message.channel.send("❌ This command can only be used in a server")
        return

    if not args:
        await message.channel.send("❌ Usage: `!ytp <number>`")
        return

    try:
        num = int(args[0])
    except ValueError:
        await message.channel.send("❌ Usage: `!ytp <number>`")
        return

    user_id = str(message.author.id)
    cached = bot.yt_search_sessions.get(user_id, [])
    if not cached or num < 1 or num > len(cached):
        await message.channel.send("❌ Invalid selection. Use `!yts` to search first")
        return

    item = cached[num - 1]
    url = item.get("url") or f"https://youtube.com/watch?v={item.get('id', '')}"

    voice_channel = get_voice_channel(message)
    if not voice_channel:
        await message.channel.send("❌ You must be in a voice channel to use this command")
        return

    status_msg = await message.channel.send(f"🔍 Loading: {item.get('title', 'video')}...")

    result = await ipc.youtube_play(
        guild_id=str(message.guild.id),
        channel_id=str(voice_channel.id),
        url=url,
        user_id=str(message.author.id),
    )

    if result.get("success"):
        title = result.get("title", item.get("title", "YouTube Video"))
        duration = result.get("duration")
        dur_str = format_duration(duration) if duration else "Live/Unknown"
        await status_msg.edit(content=f"📺 **Now Streaming:** {title}\n⏱️ Duration: {dur_str}\n🎬 Enjoy the show!")
    else:
        await status_msg.edit(content=f"❌ {result.get('error', 'Failed to play')}")
